DICOM/src/Microsoft.Health.DeID.SharedLib/CryptoHashFunction.cs

```diff
@@ -0,0 +1,107 @@
+            {
+                case CryptoHashType.HMACSHA1:
+                    _hmac = byteKey == null ? new HMACSHA1() : new HMACSHA1(byteKey);
+                    break;
+                case CryptoHashType.HMACSHA256:
+                    _hmac = byteKey == null ? new HMACSHA256() : new HMACSHA256(byteKey);
+                    break;
+                case CryptoHashType.HMACSHA512:
+                    _hmac = byteKey == null ? new HMACSHA512() : new HMACSHA512(byteKey);
+                    break;
+                default:
+                    break;
+            }
+        }
+
+        public byte[] ComputeHmacHash(byte[] input)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return _hmac.ComputeHash(input);
+        }
+
+        public byte[] ComputeHmacHash(Stream input)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            return _hmac.ComputeHash(input);
+        }
+
+        public byte[] ComputeHmacHash(string input, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            encoding ??= Encoding.UTF8;
+            return ComputeHmacHash(encoding.GetBytes(input));
+        }
+
+        public FixedLengthString ComputeHmacHash(FixedLengthString input, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+
+            var hashData = ComputeHmacHash(input.ToString(), encoding);
+
+            return new FixedLengthString(input.GetLength(), string.Concat(hashData.Select(b => b.ToString("x2"))));
+        }
+
+        public byte[] ComputeHmacHash(byte[] input, HMAC hashAlgorithm)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            return hashAlgorithm.ComputeHash(input);
+        }
+
+        public byte[] ComputeHmacHash(string input, HMAC hashAlgorithm, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            encoding ??= Encoding.UTF8;
+            return hashAlgorithm.ComputeHash(encoding.GetBytes(input));
+        }
+
+        public byte[] ComputeHmacHash(Stream input, HMAC hashAlgorithm)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            return hashAlgorithm.ComputeHash(input);
+        }
+
+        public FixedLengthString ComputeHmacHash(FixedLengthString input, HMAC hashAlgorithm, Encoding encoding = null)
+        {
+            EnsureArg.IsNotNull(input, nameof(input));
+            EnsureArg.IsNotNull(hashAlgorithm, nameof(hashAlgorithm));
+
+            var hashData = ComputeHmacHash(input.ToString(), hashAlgorithm, encoding);
+
+            input.SetString(string.Concat(hashData.Select(b => b.ToString("x2"))));
```

Comment:
❓ [Question]
Why is the behavior here different from previous `ComputeHmacHash(FixedLengthString input, Encoding encoding)`?
In that function, a new FixedLengthString is returned.
In this function, result is put into the input argument.
🐞  or by design?

Reply:
Fixed it.